File: coffeetime-parent/RubyGun/src/22_module.rb
# require
require './utils'
require './utils'

StrUtils.echo
puts StrUtils::COUNT

# load './utils.rb'
# load './utils.rb'
# StrUtils.echo

# include and included
include StrUtils
# puts echo()
sentents = ['z', 'b', 'e', 'p']
puts StrUtils.sort_words(sentents)
puts sort_words(sentents)

# inheritance
module A
  def a1
    puts 'a1'
  end

  def a2
    puts 'a2'
  end

  def self.included(base)
    puts "A is included in #{base}"
  end
end
module B
  def b1
    puts 'b1'
  end

  def b2
    puts 'b2'
  end
end
module C
  def c1
    puts 'c1'
  end

  def self.extended(base)
    puts "A is extended in #{base}"
  end
end

class Sample
  include A
  include B
  extend C

  def s1
    puts 's1'
  end
end

samp = Sample.new
samp.a1
samp.a2
samp.b1
samp.b2
samp.s1
Sample.c1

samp.extend C
samp.c1

File: coffeetime-parent/Python3Gun/src/22_module.py
# import
import utils

utils.StrUtils.echo()  # 10 20
utils.StrUtils.cout()  # 10 20
print(utils.StrUtils.COUNT)  # 20
print(utils.COUNT)  # 10

# from import
from utils import StrUtils

sentents = ['z', 'b', 'e', 'p']
print(StrUtils.sort_words(sentents))

# alias
import utils as helper

print(helper.COUNT)

# module searching path
import sys

print(sys.path)

print(__name__)  # __main__
print(helper.__name__)  # utils

from foopkg.calculator import add

print(add(3, 4))


#
# # 一个模块就是一个py文件
# # 模块的搜索路径
# import sys
# print(sys.path)
#
# # package
# # 文件夹名即包名，创建包
# # 在文件加下创建 __init__.py 文件，内容可以为空
